palace/utils/geodata.cpp

```diff
@@ -993,95 +992,191 @@ BoundingBox BoundingBoxFromPointCloud(MPI_Comm comm,
-        vertices.begin(), vertices.end(),
-        [&delta, PerpendicularDistance](const auto &x, const auto &y)
-        { return PerpendicularDistance({delta}, x) < PerpendicularDistance({delta}, y); });
-    constexpr double rel_tol = 1.0e-6;
-    MFEM_VERIFY(std::abs(PerpendicularDistance({delta}, perp) - ball.radius) <=
-                    rel_tol * ball.radius,
-                "Furthest point perpendicular must be on the exterior of the ball: "
-                    << PerpendicularDistance({delta}, perp) << " vs. " << ball.radius
-                    << "!");
-
-    // Compute a perpendicular to the circle using the cross product.
-    const Eigen::Vector3d n_radial = (perp - CVector3dMap(ball.center.data())).normalized();
-    Vector3dMap(ball.planar_normal.data()) = delta.cross(n_radial).normalized();
-
-    // Compute the point furthest out of the plane discovered. If below tolerance, this
-    // means the ball is 2D.
-    const auto &out_of_plane = *std::max_element(
-        vertices.begin(), vertices.end(),
-        [&delta, &n_radial, PerpendicularDistance](const auto &x, const auto &y)
-        {
-          return PerpendicularDistance({delta, n_radial}, x) <
-                 PerpendicularDistance({delta, n_radial}, y);
-        });
+  // If p is outside the sphere, recurse for (P \ {p}, R U {p}).
+  constexpr double rel_tol = 1.0e-6;
+  if ((vertices[p] - D.origin).norm() >= (1.0 + rel_tol) * D.radius)
+  {
+    R.push_back(p);
+    D = Welzl(P, R, vertices);
+  }
 
-    ball.planar =
-        PerpendicularDistance({delta, n_radial}, out_of_plane) / ball.radius < rel_tol;
-    if (!ball.planar)
-    {
-      // The points are not functionally coplanar, zero out the normal.
-      MFEM_VERIFY(std::abs(PerpendicularDistance({delta}, perp) - ball.radius) <=
-                      rel_tol * ball.radius,
-                  "Furthest point perpendicular must be on the exterior of the sphere!");
-      Vector3dMap(ball.planar_normal.data()) *= 0;
-    }
+  return D;
+}
+
+// Calculates a bounding ball from a point cloud using Welzl's algorithm, result is
+// broadcast across all processes. We don't operate on the convex hull, since the number of
+// points should be small enough that operating on the full set should be OK. If only three
+// points are provided, the bounding circle is computed (likewise for if the points are
+// coplanar).
+BoundingBox BoundingBallFromPointCloud(MPI_Comm comm,
+                                       const std::vector<Eigen::Vector3d> &vertices,
+                                       int dominant_rank)
+{
+  BoundingBox ball;
+  if (dominant_rank == Mpi::Rank(comm))
+  {
+    // Randomly permute the point set.
+    MFEM_VERIFY(vertices.size() >= 3,
+                "A bounding ball requires a minimum of three vertices for this algorithm!");
+    std::vector<std::size_t> indices(vertices.size() + 4);
+    for (std::size_t i = 0; i < vertices.size(); i++)
+    {
+      indices[i] = i;
+    }
```

Comment:
Alternative: size exactly then `iter_swap`, `p_i` to the end and `std::iota(vertices.begin(), vertices.end(), indices.begin(), 0);`

Reply:
Done in [32ebd0c](https://github.com/awslabs/palace/pull/246/commits/32ebd0c207e95e6a463007517689f75af0240dd1).